20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,558 @@
+
+For example,
+
+- parents of a document have to be objects of type folder
+- owners of a repository have to be of type user or a userset of group members
+
+We will add an `directly_related_user_types` to each relation to indicate what types of users can be directly related to it.
+
+`directly_related_user_types` will be an array of objects, each object must have a type and an optional relation.
+
+- having the following `"parent": { "directly_related_user_types": [{ "type": "group" }] }` in the group type definition indicates that only objects of type group can be directly related to a group as parent
+- having the following `"member": { "directly_related_user_types": [{ "type": "user" }, { "type": "group", "relation": "member" }] }` in the group type definition indicates that only objects of type user or usersets of type group and relation member can be directly related to a group as member
+- having the following `"member": { "directly_related_user_types": [{ "type": "user", "any": true }, { "type": "employee", "any": true }] }` in the group type definition indicates that the `*` syntax is allowed and that when present, it means all objects of type `user` or `employee` can be directly related to a `group` as `member`
+
+This will affect only relations that are [directly related](https://openfga.dev/docs/modeling/building-blocks/direct-relationships) (as in they are considered "assignable" and have [the direct relationship keyword ("this")](https://openfga.dev/docs/configuration-language#the-direct-relationship-keyword) in their relation definition).
+For relation definitions that:
+
+- have the direct relationship keyword (this): `directly_related_user_types` must be present and MUST be an non-empty array containing at least one type/relation combination
+- do not have the direct relationship keyword: `directly_related_user_types` can optionally be present, but if it does MUST be an empty array
+
+> Note: This syntax does not offer a way to enforce restrictions based on what the userset of group members resolves to (for example, group member can be a user, an employee or another userset). Later on, tooling can help visually indicate this to the user inline.
+
+To prevent breaking changes, this will be done by appending a new field called `metadata` -> `relations` to each type definition. Each relation will be a key under this that is a map that contains a field called `directly_related_user_types` that contains what type or type and relation combination could be directly related to it. If the `*` element is valid for that relation, then the type and `any: true` need to be in the `directly_related_user_types`. If multiple entries that contains `any: true` are present, a `*` in a relationship tuple will be interpreted to mean any user of the union of those types.
```

Comment:
```suggestion
To prevent breaking changes, this will be done by appending a new field called `metadata` -> `relations` to each type definition. Each relation will be a key under this that is a map that contains a field called `directly_related_user_types` that contains what type or type and relation combination could be directly related to it. If the `*` element is valid for that relation, then the type and `any: true` needs to be in `directly_related_user_types`. If multiple entries where `any: true` are present, a `*` in a relationship tuple will be interpreted to mean any user of the union of those types.
```

But I am not sure I understand the last sentence. An example could help.

Reply:
Added an example